apiserver/migrationtarget/migrationtarget.go

```diff
@@ -223,3 +224,70 @@ func (api *API) AdoptResources(args params.AdoptResourcesArgs) error {
 	}
 	return errors.Trace(env.AdoptResources(model.ControllerUUID(), args.SourceControllerVersion))
 }
+
+// CheckMachines compares the machines in state with the ones reported
+// by the provider and reports any discrepancies.
+func (api *API) CheckMachines(args params.ModelArgs) (params.ErrorResults, error) {
+	var empty params.ErrorResults
+	model, err := api.getModel(args.ModelTag)
+	if err != nil {
+		return empty, errors.Trace(err)
+	}
+	st, release, err := api.pool.Get(model.UUID())
+	if err != nil {
+		return empty, errors.Trace(err)
+	}
+	defer release()
+
+	machines, err := st.AllMachines()
+	if err != nil {
+		return empty, errors.Trace(err)
+	}
+	machinesByInstance := make(map[string]string)
+	for _, machine := range machines {
+		if machine.IsContainer() {
+			// Containers don't correspond to instances at the
+			// provider level.
+			continue
+		}
+		instanceId, err := machine.InstanceId()
+		if err != nil {
+			return empty, errors.Annotatef(
+				err, "getting instance id for machine %s", machine.Id())
+		}
+		machinesByInstance[string(instanceId)] = machine.Id()
+	}
+
+	env, err := api.getEnviron(st)
+	if err != nil {
+		return empty, errors.Trace(err)
+	}
+	instances, err := env.AllInstances()
+	if err != nil {
+		return empty, errors.Trace(err)
+	}
+
+	var results []params.ErrorResult
+
+	instanceIds := set.NewStrings()
+	for _, instance := range instances {
+		id := string(instance.Id())
+		instanceIds.Add(id)
+		if _, found := machinesByInstance[id]; !found {
+			results = append(results, errorResult("no machine with instance %q", id))
+		}
+	}
+
+	for instanceId, name := range machinesByInstance {
+		if !instanceIds.Contains(instanceId) {
+			results = append(results, errorResult(
+				"couldn't find instance %q for machine %s", instanceId, name))
+		}
+	}
+
+	return params.ErrorResults{Results: results}, nil
```

Comment:
This is a slightly unusual use of `ErrorResults`. Typically each element in the `Results` field corresponds to the same position in an input list but here it's being used to return an arbitrary number of errors. I guess that's ok though.